### ISSUE TITLE
rsyslogd leak memory when sending logs to rsyslogd  frequently about 10000 line per second

### ISSUE THREAD

**Issue by john1jxm** [rsyslog/rsyslog#4759]
**at beginning, used memory is small, 1988KB**
```
[root@Beacon 1.1: /logs]# ps |grep syslog
30020 root     19708 S    rsyslogd -f /tmp/rsyslog.conf
[root@Beacon 1.1: /logs]# cat /proc/30020/status |grep VmRSS
VmRSS:      1988 kB
```
**send logs to rsyslogd frequently about 1000 line per second,**
the used peak memory is 33436KB
[root@Beacon 1.1: /logs]# cat /proc/30020/status |grep VmRSS
VmRSS:     33436 kB

**stop to send logs  , 20 minutes later, check memory, the used memory is still 15152KB which is larger then 1988KB** 
[root@Beacon 1.1: /logs]# cat /proc/30020/status |grep VmRSS
VmRSS:     15152 kB
is there memory leak?

after send 
### Expected behavior
after stop to send logs to rsyslogd, the used memory of rsysogd should be freed

### Actual behavior
no free.
### Steps to reproduce the behavior
rsyslogd -f /tmp/rsyslog.conf

### Environment
- rsyslog version: 8.2104.0
- platform:  mips GNU/Linux linux
- for configuration questions/issues, include rsyslog.conf and included config files
$ModLoad imuxsock.so
$ModLoad imklog.so
$template EventMsgFormat,"%timegenerated%[%syslogseverity-text%]%msg%\n"
$template FileFormat,"[%syslogseverity-text%] <%PRI%>1 %TIMESTAMP:::date-rfc3339% %HOSTNAME% %APP-NAME% %PROCID% %MSGID% %STRUCTURED-DATA% %msg%\n"
$RuleSet local
$outchannel messages_rotation,/logs/messages, 2097152,/usr/exe/syslog_rotation_script.sh /logs/messages
$outchannel customer_rotation,/logs/customer, 2097152,/usr/exe/syslog_rotation_script.sh /logs/customer
$outchannel srmf_abnormal_log_rotation,/logs/srmf_abnormal_log, 2097152,/usr/exe/syslog_rotation_script.sh /logs/srmf_abnormal_log
$outchannel srmf_normal_log_rotation,/logs/srmf_normal_log, 2097152,/usr/exe/syslog_rotation_script.sh /logs/srmf_normal_log
auth,authpriv,user.notice;local0.notice;kern,syslog,daemon.notice :omfile:$messages_rotation;FileFormat
local1.*  |/tmp/event_messages;EventMsgFormat
local0.notice :omfile:$customer_rotation;FileFormat
local2.=info :omfile:$srmf_normal_log_rotation;FileFormat
local2.=alert :omfile:$srmf_abnormal_log_rotation;FileFormat
$DefaultRuleset local

**Comment by davidelang** [comment]
please enable impstats and show us the full config.

David Lang

**Comment by john1jxm** [comment]
how to enable impstats

**Comment by davidelang** [comment]
https://www.rsyslog.com/doc/v8-stable/configuration/modules/impstats.html
https://www.rsyslog.com/how-to-use-impstats/

David Lang

**Comment by john1jxm** [comment]
root@Cortina:/logs# cat /tmp/stats.log
Fri May 22 03:33:11 2020: global: origin=dynstats
Fri May 22 03:33:11 2020: imuxsock: origin=imuxsock submitted=3332711 ratelimit.discarded=0 ratelimit.numratelimiters=0
Fri May 22 03:33:11 2020: action-0-builtin:omfile: origin=core.action processed=3319150 failed=0 suspended=0 suspended.duration=0 resumed=0
Fri May 22 03:33:11 2020: action-1-builtin:omfile: origin=core.action processed=0 failed=0 suspended=0 suspended.duration=0 resumed=0
Fri May 22 03:33:11 2020: resource-usage: origin=impstats utime=55294686 stime=27597937 maxrss=64260 minflt=879642 majflt=0 inblock=0 oublock=0 nvcsw=341980 nivcsw=116098 openfiles=5
Fri May 22 03:33:11 2020: main Q: origin=core.queue size=0 enqueued=3333485 full=25 discarded.full=0 discarded.nf=0 maxqsize=100000
root@Cortina:/logs#
root@Cortina:/logs#
root@Cortina:/logs# cat /tmp/rsyslog.conf
module(load="imuxsock")
module(load="imklog")
module(load="impstats"
       interval="600"
       severity="7"
       log.syslog="off"
       log.file="/tmp/stats.log")
$template FileFormat,"[%syslogseverity-text%] <%PRI%>1 %TIMESTAMP:::date-rfc3339% %HOSTNAME% %APP-NAME% %PROCID% %MSGID% %STRUCTURED-DATA% %msg%\n"
$outchannel messages_rotation,/logs/messages, 2097152,/usr/sbin/syslog_rotation_script.sh /logs/messages
user.notice;kern,syslog,daemon.notice :omfile:$messages_rotation;FileFormat
$outchannel customer_rotation,/logs/customer, 2097152,/usr/sbin/syslog_rotation_script.sh /logs/customer
auth,authpriv.*;local0.notice :omfile:$customer_rotation;FileFormat
root@Cortina:/logs#

**Comment by rgerhards** [comment]
I would tend to say that this is buffer memory (libc allocator keeps some buffers for reuse). What happens if you run steady traffic for a couple of hours. Does memory continuously go up?

**Comment by john1jxm** [comment]
yes, it continuously go up, but my quesion is  that why the momory didn't drop down. it keep as 15152 kB.
I don't think it's glibc memory buffer. for my embedded device,  15152 kB memory is a big.

**Comment by john1jxm** [comment]
Is there any parameters to limit the buffer memory, for the embedded device,  the memory useage is sensitive.

**Comment by rgerhards** [comment]
I re-triaged this while processing the old crash/memory issue queue.

Current conclusion: I do not see enough evidence here to make a safe code change or to close the issue.

What the existing data shows:

- The impstats sample had `main Q` size `0`, so the sample does not show messages stuck in the main queue at that point.
- The RSS staying above the initial process size after a high-throughput burst can still be allocator/cache behavior, especially on older libc/platform combinations.
- The report is from rsyslog 8.2104.0 on MIPS and does not include a Valgrind leak trace, heap profile, or current-version reproduction.

To move this from “possible memory retention” to an actionable rsyslog leak, we would need a current rsyslog version test plus one of:

- a Valgrind/memcheck leak log from shutdown,
- a heap profile showing the growing allocation site,
- or repeated impstats samples during steady traffic showing which queue/action/input grows without returning.

I am leaving this open rather than closing it, because the report is under-specified and may still matter for embedded deployments, but it is not actionable as a code change from the data currently attached.
